### DISCUSSION TITLE
IPC improvements

### DISCUSSION THREAD

**Discussion by ep12** [qtile/qtile#2337] [Ideas]
Summary over at #2338:

This discussion came about when fixing `cmd_display_kb` for `KeyChords`, but it seems there might be a need to discuss how the IPC works vs. how it should work. I haven't yet had a look at it, but an idea I had a while ago was to send and receive `pickle` objects, because we might want to get information in a structured format, not as a string.

This is just an idea to start the discussion, so feel free to express your ipc-related thoughts, including things you find annoying, you want to keep or you want to change.

**Comment by ramnes** [comment]
On `display_kb` specifically, I think what we all want is both an easy to parse output and a nice helper for everyday use. At the moment, it's neither one nor the other, because it has been made for `qtile shell`, which is fun but not really practical.

To achieve this, we could probably 1/ add a `cmd_keys` function similar to `cmd_windows`, that just prints the Python list of keybindings
and 2/ deprecate `display_kb` and replace it with another command that it spawns some kind of window (a popup? a terminal? other?) containing nicely formatted helper.

**Reply by ep12** [reply]
Let's mention #1329 here :)

**Comment by ramnes** [comment]
On our interactions with the qtile process, I think part of the problem is that we wrongly call them IPC. We're not building a microservices architecture here. What we need is a simple API, that can be both read and parsed easily, whatever the programming language. Funny thing is, that's exactly what JSON was invented for, so we should probably use it. :)

**Reply by flacjacket** [reply]
> On our interactions with the qtile process, I think part of the problem is that we wrongly call them IPC.

I'm curious why you wouldn't classify this as IPC? I agree, this isn't microservices, but we are still passing data from the qtile shell process into the running qtile process and back.

**Reply by flacjacket** [reply]
> Funny thing is, that's exactly what JSON was invented for, so we should probably use it. :)

This is definitely an interesting idea. I know we have the json functionality in the current IPC clients to make it easier to return data to other 3rd party clients that don't want to marshal data in and out. Where that could be a problem is if there is any place where we are sending objects rather than built-in types, but I'm not aware of any instances of that in our current code base. The other tradeoff would be that something like pickle or marshal should be less data and would be (un)serialized faster, but I would doubt that would be a critical bottleneck for us.

**Reply by ramnes** [reply]
Ahah yeah, `qtile shell` is probably the only "real IPC" thing we're doing. :)

Glad to see you around!

**Comment by ep12** [comment]
+ `_IPC`: `is_json` defaults to `None`, meaning that unpacking a string as `json` is attempted. `Client`: `is_json` defaults to `False`? Is there a reason for that?
+ We are using `marshal`. The [python docs](https://docs.python.org/3/library/marshal.html) say
  > If you’re serializing and de-serializing Python objects, use the pickle module instead – the performance is comparable, version independence is guaranteed, and pickle supports a substantially wider range of objects than marshal.
+ Why don't we pickle by default? JSON-compatible list-dict-string-number object are a subset of all `pickle`able objects?
+ Pickling custom objects is possible, JSON cannot do that: https://docs.python.org/3/library/pickle.html?highlight=pickle#comparison-with-json
+ (Pickling can be combined with signing if decided to be required: https://docs.python.org/3/library/pickle.html?highlight=hmac#module-pickle)
+ The command interface consists of a lot of classes. Can that be boiled down? Just a question to which I currently don't have an answer as I am not too familiar with the current ipc.

**Reply by flacjacket** [reply]
>     * `_IPC`: `is_json` defaults to `None`, meaning that unpacking a string as `json` is attempted. `Client`: `is_json` defaults to `False`? Is there a reason for that?

I think this is also for the sake of the server. On the server side, you don't know if the message is json or marshaled, so you need to be able to check both. When you are the client, you can pick either json or marshal, and you expect the message to come back with the same encoding, so you don't need that fallback to un-marshaling when you expect json, and you don't need to try json when you expect marshaled bytes.

>     * We are using `marshal`. The [python docs](https://docs.python.org/3/library/marshal.html) say
>       > If you’re serializing and de-serializing Python objects, use the pickle module instead – the performance is comparable, version independence is guaranteed, and pickle supports a substantially wider range of objects than marshal.
> 
>     * Why don't we pickle by default? JSON-compatible list-dict-string-number object are a subset of all `pickle`able objects?
> 
>     * Pickling custom objects is possible, JSON cannot do that: https://docs.python.org/3/library/pickle.html?highlight=pickle#comparison-with-json
> 
>     * (Pickling can be combined with signing if decided to be required: https://docs.python.org/3/library/pickle.html?highlight=hmac#module-pickle)

These are great questions. The answer may lie back in the _very first_ qtile commit by Aldo Cortesi back in 2008 ec7f8b15c72ebccb53d9ba8c934a5828a577e209, where he chose marshal, and we haven't come back to update it since :)

>     * The command interface consists of a lot of classes. Can that be boiled down? Just a question to which I currently don't have an answer as I am not too familiar with the current ipc.

This may be a bit of my doing, so I'm happy to explain the solution that we wound up at and the different components, and also help figure out what can be done to improve it. My attempt to capture this in the docs starts roughly here: http://docs.qtile.org/en/latest/manual/commands/index.html#digging-deeper-command-objects

Taking another run at trying to break down the different components, in case this helps frame it in a different light:

The backbone of the command interface is the abstract graph. This is essentially a collection of classes that let you traverse through the set legal [0] object that you may issue a command against and formulate a path to the object along with the command and the arguments to pass into it. This helps, for example, the `qtile shell` to parse the command that is issued and represent this as a traversal of this abstract graph without needing to make any calls to the running qtile instance, or allow the lazy object to use the same graph traversal and execution mechanism (the clients) as is used when executing commands on the running qtile instance.

Once you have a command that you have defined in this graph space, you need to be able to have some way to execute the command. This is what is handled by the `interface`'s. These take one of the graph call objects, execute it, and return the result. How this happens is different if you are running in the active qtile process (you can directly invoke the methods on the objects), if you are running from a separate process e.g. via `qtile shell` (you need to invoke the IPC to send the command and receive the result), or you are defining a `lazy` call (you just return a `LazyCall` object). The interface, since it represents the real qtile instance, also lets you look and see if an object actually exists in a graph. This is useful for doing the traversal and knowing what objects are valid (eg if you are in the `shell` and `ls` the windows on the screen), or in the case of the Lazy objects, allowing all traversals to be allowed.

The final piece is the different clients. These are basically the objects that do the abstract graph traversal and then execute the command against the command interface. Essentially, the thing that combines the functionality together. There are 2 main ones here: the interactive command client is what lets you do things like `client.window.dox()`, and it resolves where that is in the graph and dispatches the `cmd_dox` on the current `window`, and it uses the given interface to do it. This can be used by the `lazy` object, for example, to be able to make this sort of call and build up a lazy object, or it can be used by the qtile shell to allow for the user to issue commands using the IPC interface. The other client that currently exists is the command client. This seems to not have gotten a lot of use, or maybe the uses for it that once existed were removed, or the migration to use this was never done. The idea with the command client is this is a more programatic way to traverse the graph. There are currently a few places where we just `eval("...")` to be able to invoke the interactive client or do things like call `getattr` on the interactive client to resolve some path, where ideally, these could be changed over to more explicit navigating through the graph with the command client.

Overall, this can be summarized as: there is a `graph` structure that qtile defines between its objects and commands, so we define a common set the nodes and connections in that graph. The user wants to be able to traverse and execute commands on the graph, which is handled by the `client`. There are several implementations of how we execute those commands, and those differences are captured in the different `interfaces`, which is what allows the `client` the ability to actually run the commands. The current implementation was arrived at by trying to separate out these different components so we could leverage the same functionality between the various use cases for making calls (in-process, out-of-process, lazy) and be able to more easily represent locations and calls on objects in the graph without needing to use the live qtile objects themselves.

[0] There are baked in here assumptions about what components hold what other components. This is maybe baking a bit of the implementation into what is called as abstract call graph, but helps us to know when we say `screen.window`, we want a window object on the current screen, not the `cmd_window` on the current screen object.

**Reply by ep12** [reply]
Thank you for this very detailed explanation! This is definitely helpful to get all those who participate in the discussion to (or towards) the same level.
I already knew that the graph was the heart of it, but now I know a lot more about the different classes that are involved.
What I still need to do is to have a look at the lazy stuff to understand it and see if a change to pickle or JSON results in problems there (hopefully not).

**Reply by flacjacket** [reply]
The lazy components should be totally separate from the serialization logic. The serialization only occurs when you are using the `IPCCommandInterface` to execute the graph calls. The `IPCCommandInterface` invokes the underlying `ipc.Client`, which is what handles the IPC to and from the qtile instance. For the `lazy` object, the `LazyCommandInterface` is used, which is just a stub that allows the users to navigate the graph and store the resulting call in the `LazyCall` for execution later. We should be totally free to change the serialization without affecting anything on the `lazy` side.
